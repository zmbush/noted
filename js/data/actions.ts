// Copyright 2019 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

import { NoteData, ErrorData, UserData } from 'data/types';
import axios from 'axios';
import { Dispatch } from 'redux';

export enum NotedEvent {
  NotesFetchStart = 'NOTES_FETCH_START',
  NotesFetched = 'NOTES_FETCHED',
  NotesUpdateNote = 'NOTES_UPDATE_NOTE',
  NotesDeleteNote = 'NOTES_DELETE_NOTE',

  ApiError = 'API_ERROR',

  UserSignedIn = 'USER_SIGNED_IN',
  UserSignedOut = 'USER_SIGNED_OUT',
}

export function notesFetchStart() {
  return { type: NotedEvent.NotesFetchStart };
}

export function notesFetched(notes: NoteData[]) {
  return { type: NotedEvent.NotesFetched, notes };
}

export function updateNote(note: NoteData) {
  return { type: NotedEvent.NotesUpdateNote, note };
}

export function deleteNote(id: number) {
  return { type: NotedEvent.NotesDeleteNote, id };
}

export function apiError(error: ErrorData) {
  return { type: NotedEvent.ApiError, error };
}

export function logIn(user: UserData) {
  return { type: NotedEvent.UserSignedIn, user };
}

export function logOut() {
  return { type: NotedEvent.UserSignedOut };
}

export async function fetchData(dispatch: Dispatch) {
  dispatch(notesFetchStart());
  try {
    dispatch(notesFetched((await axios.get('/api/secure/notes')).data));
  } catch (e) {
    if (e.response) {
      dispatch(apiError(e.response.data));
    }
  }
}
