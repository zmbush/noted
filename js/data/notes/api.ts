// Copyright 2021 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.
//
import { createAsyncThunk } from '@reduxjs/toolkit';

import api from 'api';

import { NewNote, NoteWithTags, UpdateNote } from '../types';

export const getNotes = createAsyncThunk('notes/getList', (_: void) => api.note.list());
export const updateNote = createAsyncThunk(
  'notes/updateNote',
  async ({
    noteId,
    note: noteIn,
  }: {
    noteId: number;
    note: UpdateNote & Partial<Pick<NoteWithTags, 'tags'>>;
  }) => {
    const { tags, ...note } = noteIn;
    let result = await api.note.update(noteId, note);
    if (tags) {
      result = await api.note.setTags(result.id, tags);
    }
    return result;
  },
);
export const createNote = createAsyncThunk(
  'notes/createNote',
  async (noteIn: NewNote & Pick<NoteWithTags, 'tags'>) => {
    const { tags, ...note } = noteIn;
    const result = await api.note.create(note);
    return api.note.setTags(result.id, tags);
  },
);
export const deleteNote = createAsyncThunk('notes/deleteNote', async (noteId: number) => {
  await api.note.delete(noteId);
  return noteId;
});
