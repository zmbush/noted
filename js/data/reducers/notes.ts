// Copyright 2019 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

import { NotedEvent } from 'data/actions';
import { NoteData, ErrorData } from 'data/types';
import update from 'immutability-helper';

const initialState = new Map<number, NoteData>();
type State = typeof initialState;

export default function notes(
  state = initialState,
  action: {
    type?: NotedEvent;
    notes?: NoteData[];
    note?: NoteData;
    error?: ErrorData;
    id?: number;
  }
): State {
  switch (action.type) {
    case NotedEvent.NotesFetched: {
      let data = new Map();
      for (let note of action.notes) {
        data.set(note.id, note);
      }
      return data;
    }
    case NotedEvent.NotesUpdateNote: {
      let data = new Map(state);
      data.set(action.note.id, action.note);
      return data;
    }
    case NotedEvent.NotesDeleteNote: {
      let data = new Map(state);
      data.delete(action.id);
      return data;
    }
    case NotedEvent.ApiError: {
      if (action.error.code == 401) {
        return new Map();
      }
    }
    case NotedEvent.UserSignedOut: {
      return new Map();
    }
    default:
      return state;
  }
}
