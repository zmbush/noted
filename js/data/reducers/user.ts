// Copyright 2019 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.
import update from 'immutability-helper';

import { NotedEvent } from 'data/actions';
import { User, ErrorData } from 'data/api_types';

const initialState = { is_signed_in: false, user: null as null | User };

type State = typeof initialState;

export default function user(
  // eslint-disable-next-line default-param-last
  state = initialState,
  action: {
    type?: NotedEvent;
    error?: ErrorData;
    is_signed_in?: boolean;
    user?: User;
  },
): State {
  switch (action.type) {
    case NotedEvent.ApiError: {
      if (action.error.code === 401) {
        return initialState;
      }
      return state;
    }
    case NotedEvent.UserSignedIn: {
      return update(state, {
        is_signed_in: { $set: true },
        user: { $set: action.user },
      });
    }
    case NotedEvent.UserSignedOut: {
      return initialState;
    }
    default:
      return state;
  }
}
