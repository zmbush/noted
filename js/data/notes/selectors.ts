// Copyright 2019 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.
//
import { EntityState } from '@reduxjs/toolkit';
import createCachedSelector from 're-reselect';
import { createSelector } from 'reselect';

import { AppState } from 'data/store';
import { NoteWithTags } from 'data/types';

import { prefix } from './api';
import { notesAdapter } from './slice';

export const getNotes = (state: AppState) => state[prefix];
const noteSelectors = notesAdapter.getSelectors(getNotes);
export const getNoteEntities = createSelector(noteSelectors.selectEntities, (notesDict) =>
  Object.fromEntries(Object.entries(notesDict).map(([key, note]) => [key, note!])),
);
// const getNoteId = (_: any, props: { noteId: number }) => props.noteId;
const maybeGetNoteId = (_: any, props: { noteId: number | null }) => props.noteId;
const maybeNoteIdIndex = (_: any, props: { noteId: number | null }) =>
  props.noteId === null ? -1 : props.noteId;
export const listNotes = noteSelectors.selectAll;
const validParent = (note: NoteWithTags) => !!note.parent_note_id;

const listAllTitles = createSelector(listNotes, (notes) =>
  notes.map((note) => [note.id, note.title]),
);

const listAllTitleParts = createSelector(listAllTitles, (titleList: [number, string][]) =>
  titleList.flatMap(([id, title]) => {
    const titles = [[title, id]];
    title.split(' ').forEach((titlePart) => {
      if (titlePart.length > 3) {
        titles.push([titlePart, id]);
      }
    });
    return titles;
  }),
);

export const getLinkIds = createSelector(listAllTitleParts, (titleParts: [string, number][]) =>
  titleParts.reduce((titles, [titlePart, id]) => {
    if (titlePart in titles) {
      titles[titlePart].add(id);
    } else {
      // eslint-disable-next-line no-param-reassign
      titles[titlePart] = new Set([id]);
    }
    return titles;
  }, {} as { [id: string]: Set<number> }),
);

export type LinkIdMap = ReturnType<typeof getLinkIds>;

// export const getTopLevelNotes = createSelector(listNotes, (notes) => {
//   const topLevel: { [id: number]: NoteWithTags } = {};
//   notes.forEach((note) => {
//     if (!validParent(note)) {
//       topLevel[note.id] = note;
//     }
//   });
//   return topLevel;
// });

export const getSubNotes = createCachedSelector(listNotes, maybeGetNoteId, (notes, noteId) => {
  const subNotes: { [id: number]: NoteWithTags } = {};

  if (!noteId) {
    notes.forEach((note) => {
      if (!validParent(note)) {
        subNotes[note.id] = note;
      }
    });
  } else {
    notes.forEach((note) => {
      if (note.parent_note_id === noteId && note.id !== noteId) {
        subNotes[note.id] = note;
      }
    });
  }

  return subNotes;
})(maybeNoteIdIndex);

export type SubNoteMap = ReturnType<typeof getSubNotes>;

export const getIsNotArchived = createSelector(getNoteEntities, (notes) => {
  const isNotArchived: { [id: number]: boolean } = {};
  Object.values(notes).forEach((note: NoteWithTags) => {
    isNotArchived[note.id] = !note.archived;
  });
  return isNotArchived;
});

export const getHasArchivedChild = createSelector(getNoteEntities, (notes) => {
  const hasArchivedChild: { [id: number]: boolean } = {};
  Object.values(notes).forEach((note: NoteWithTags) => {
    const arch = note.archived;
    hasArchivedChild[note.id] = hasArchivedChild[note.id] || arch;
    let currentNote = note;
    while (validParent(currentNote)) {
      currentNote = notes[currentNote.parent_note_id]!;
      hasArchivedChild[currentNote.id] = hasArchivedChild[currentNote.id] || arch;
    }
  });
  return hasArchivedChild;
});

const calculateMergedNote = (
  note: NoteWithTags,
  notes: EntityState<NoteWithTags>,
): NoteWithTags => {
  const newNote = { ...note };
  const subNotes = getSubNotes({ notes }, { noteId: note.id });

  Object.values(subNotes).forEach((subNote) => {
    const mergedNote = calculateMergedNote(subNote, notes);
    newNote.title = `${newNote.title} ${mergedNote.title}`;
    newNote.tags = Array.from(new Set([...newNote.tags, ...mergedNote.tags]));
    newNote.body = `${newNote.body} ${mergedNote.body}`;
  });

  return newNote;
};

export const getSearchIndex = createSelector(getNotes, (allNotes) => {
  const result: { [id: number]: NoteWithTags } = {};
  Object.values(allNotes.entities).forEach((note: NoteWithTags) => {
    result[note.id] = calculateMergedNote(note, allNotes);
  });
  return result;
});

export const getFilteredSearchIndex = createCachedSelector(
  getSearchIndex,
  maybeGetNoteId,
  (notes, noteId) => {
    const subNotes: { [id: number]: NoteWithTags } = {};
    if (noteId == null) {
      // eslint-disable-next-line no-param-reassign
      noteId = 0;
    }

    Object.values(notes).forEach((note) => {
      if (note.parent_note_id === noteId) {
        subNotes[note.id] = note;
      }
    });

    return subNotes;
  },
)((state, props: { noteId: number }) => (props.noteId == null ? -1 : props.noteId));

const mostRecent = (a: string, b: string) => (a > b ? a : b);

export const getSortedNoteIds = createSelector(getNoteEntities, (allNotes) => {
  const map = new Map<number, string>();

  Object.values(allNotes).forEach((note: NoteWithTags) => {
    map.set(note.id, note.updated_at);
  });

  Object.values(allNotes).forEach((thisNote: NoteWithTags) => {
    let note = thisNote;
    while (validParent(note)) {
      map.set(
        note.parent_note_id,
        mostRecent(note.updated_at, map.get(note.parent_note_id || -1)!),
      );
      note = allNotes[note.parent_note_id]!;
    }
  });

  return Array.from(map.entries())
    .sort(([aid, a]: [number, string], [bid, b]: [number, string]) => {
      const na = allNotes[aid]!;
      const nb = allNotes[bid]!;

      if (na.pinned && !nb.pinned) {
        return -1;
      }
      if (!na.pinned && nb.pinned) {
        return 1;
      }

      if (na.archived && !nb.archived) {
        return 1;
      }
      if (!na.archived && nb.archived) {
        return -1;
      }

      if (a > b) {
        return -1;
      }
      if (a < b) {
        return 1;
      }
      return 0;
    })
    .map(([id, _]) => id);
});
