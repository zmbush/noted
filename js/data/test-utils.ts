// Copyright 2021 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.
//
import { rootReducer } from 'data/store';
import { NoteWithTags } from 'data/types';

import { getNotes } from './notes/api';

export const emptyTestNote: NoteWithTags = {
  id: -1,
  user_id: -1,
  title: 'The Note',
  body: 'The Body',
  tags: [],
  archived: false,
  pinned: false,
  parent_note_id: 0,
  created_at: '',
  updated_at: '',
};

let id = 0;
export const makeTestNote = (
  overrides: Partial<NoteWithTags> = {},
  parent: { id: number } = null,
): NoteWithTags => {
  id += 1;
  const thisId = overrides.id || id;
  const defaultFields: Partial<NoteWithTags> = {
    id: thisId,
    title: `Note ${thisId}`,
  };
  const parentOverride: Partial<NoteWithTags> = {};
  if (parent) {
    parentOverride.parent_note_id = parent.id;
  }
  return { ...emptyTestNote, ...defaultFields, ...overrides, ...parentOverride };
};

export const makeTestState = () =>
  rootReducer(
    undefined,
    getNotes.fulfilled(
      [makeTestNote(), makeTestNote({ id: 4 }), makeTestNote({ id: 2 })],
      '',
      undefined,
      undefined,
    ),
  );
