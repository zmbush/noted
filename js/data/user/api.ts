// Copyright 2021 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.
//
import { createAsyncThunk } from '@reduxjs/toolkit';

import api from 'api';
import { getNotes } from 'data/notes/api';
import { NewUserRequest, SignIn } from 'data/types';

export const getCurrentUser = createAsyncThunk(
  'users/getCurrentUser',
  async (_: void, { rejectWithValue, dispatch }) => {
    try {
      const user = await api.user.get();
      await dispatch(getNotes());
      return user;
    } catch (e) {
      throw rejectWithValue(e);
    }
  },
);

export const signInUser = createAsyncThunk(
  'user/signIn',
  async (signIn: SignIn, { rejectWithValue, dispatch }) => {
    try {
      const user = await api.user.signIn(signIn);
      await dispatch(getNotes());
      return user;
    } catch (e) {
      throw rejectWithValue(e);
    }
  },
);

export const signUpUser = createAsyncThunk(
  'user/signUp',
  async (signUp: NewUserRequest, { rejectWithValue, dispatch }) => {
    try {
      const user = await api.user.signUp(signUp);
      await dispatch(getNotes());
      return user;
    } catch (e) {
      throw rejectWithValue(e);
    }
  },
);

export const signOutUser = createAsyncThunk(
  'user/signOut',
  async (_: void, { rejectWithValue }) => {
    try {
      await api.user.signOut();
    } catch (e) {
      throw rejectWithValue(e);
    }
  },
);
