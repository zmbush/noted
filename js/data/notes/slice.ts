// Copyright 2021 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.
//
import { createEntityAdapter, createSlice } from '@reduxjs/toolkit';

import { NoteWithTags } from 'data/types';
import { signOutUser } from 'data/user/api';

import { getNotes, updateNote, createNote, deleteNote, prefix } from './api';

export const notesAdapter = createEntityAdapter<NoteWithTags>();

export const notesSlice = createSlice({
  name: prefix,
  initialState: notesAdapter.getInitialState(),
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(getNotes.fulfilled, (state, { payload }) => {
        notesAdapter.addMany(state, payload);
      })
      .addCase(updateNote.fulfilled, (state, { payload }) => {
        notesAdapter.setOne(state, payload);
      })
      .addCase(createNote.fulfilled, (state, { payload }) => {
        notesAdapter.addOne(state, payload);
      })
      .addCase(deleteNote.fulfilled, (state, { payload }) => {
        notesAdapter.removeOne(state, payload);
      })
      .addCase(signOutUser.fulfilled, (state) => {
        notesAdapter.removeAll(state);
      });
  },
});

export default notesSlice.reducer;
