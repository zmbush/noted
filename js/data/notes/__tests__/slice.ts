// Copyright 2019 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.
//
import { makeTestNote } from 'data/test-utils';
import { signOutUser } from 'data/user/api';

import { deleteNote, getNotes, updateNote } from '../api';
import notes from '../slice';

describe('reducers::notes()', () => {
  const getInitial = () => notes(undefined, { type: '' });

  test('returns initial state', () => {
    expect(getInitial()).toEqual({ entities: {}, ids: [] });
  });

  test('responds to events', () => {
    let noteOne = makeTestNote({ id: 1, title: 'title', body: 'body' });
    let noteTwo = makeTestNote({ id: 2, title: 'title 2', body: 'body' });
    let state = getInitial();

    state = notes(state, getNotes.fulfilled([noteOne], ''));
    expect(state).toMatchInlineSnapshot(`
      Object {
        "entities": Object {
          "1": Object {
            "archived": false,
            "body": "body",
            "created_at": "",
            "id": 1,
            "parent_note_id": 0,
            "pinned": false,
            "tags": Array [],
            "title": "title",
            "updated_at": "",
            "user_id": -1,
          },
        },
        "ids": Array [
          1,
        ],
      }
    `);

    noteOne = { ...noteOne, title: 'title 3' };
    state = notes(state, updateNote.fulfilled(noteOne, '', { noteId: noteOne.id, note: noteOne }));
    expect(state).toMatchInlineSnapshot(`
      Object {
        "entities": Object {
          "1": Object {
            "archived": false,
            "body": "body",
            "created_at": "",
            "id": 1,
            "parent_note_id": 0,
            "pinned": false,
            "tags": Array [],
            "title": "title 3",
            "updated_at": "",
            "user_id": -1,
          },
        },
        "ids": Array [
          1,
        ],
      }
    `);

    noteTwo = { ...noteTwo, title: 'title 4' };
    state = notes(state, updateNote.fulfilled(noteTwo, '', { noteId: noteTwo.id, note: noteTwo }));
    expect(state).toMatchInlineSnapshot(`
      Object {
        "entities": Object {
          "1": Object {
            "archived": false,
            "body": "body",
            "created_at": "",
            "id": 1,
            "parent_note_id": 0,
            "pinned": false,
            "tags": Array [],
            "title": "title 3",
            "updated_at": "",
            "user_id": -1,
          },
          "2": Object {
            "archived": false,
            "body": "body",
            "created_at": "",
            "id": 2,
            "parent_note_id": 0,
            "pinned": false,
            "tags": Array [],
            "title": "title 4",
            "updated_at": "",
            "user_id": -1,
          },
        },
        "ids": Array [
          1,
          2,
        ],
      }
    `);

    state = notes(state, deleteNote.fulfilled(noteTwo.id, '', noteTwo.id));
    expect(state).toMatchInlineSnapshot(`
      Object {
        "entities": Object {
          "1": Object {
            "archived": false,
            "body": "body",
            "created_at": "",
            "id": 1,
            "parent_note_id": 0,
            "pinned": false,
            "tags": Array [],
            "title": "title 3",
            "updated_at": "",
            "user_id": -1,
          },
        },
        "ids": Array [
          1,
        ],
      }
    `);

    state = notes(state, signOutUser.fulfilled(undefined, '', undefined));
    expect(state.ids).toHaveLength(0);
    expect(state).toMatchInlineSnapshot(`
      Object {
        "entities": Object {},
        "ids": Array [],
      }
    `);
  });
});
