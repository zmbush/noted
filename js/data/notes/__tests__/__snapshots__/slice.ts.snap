// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers::notes() responds to events 1`] = `
Object {
  "entities": Object {
    "4": Object {
      "archived": false,
      "body": "body",
      "created_at": "",
      "id": 4,
      "parent_note_id": 0,
      "pinned": false,
      "tags": Array [],
      "title": "title",
      "updated_at": "",
      "user_id": -1,
    },
  },
  "ids": Array [
    4,
  ],
}
`;

exports[`reducers::notes() responds to events 2`] = `
Object {
  "entities": Object {
    "4": Object {
      "archived": false,
      "body": "body",
      "created_at": "",
      "id": 4,
      "parent_note_id": 0,
      "pinned": false,
      "tags": Array [],
      "title": "title 3",
      "updated_at": "",
      "user_id": -1,
    },
  },
  "ids": Array [
    4,
  ],
}
`;

exports[`reducers::notes() responds to events 3`] = `
Object {
  "entities": Object {
    "4": Object {
      "archived": false,
      "body": "body",
      "created_at": "",
      "id": 4,
      "parent_note_id": 0,
      "pinned": false,
      "tags": Array [],
      "title": "title 3",
      "updated_at": "",
      "user_id": -1,
    },
    "5": Object {
      "archived": false,
      "body": "body",
      "created_at": "",
      "id": 5,
      "parent_note_id": 0,
      "pinned": false,
      "tags": Array [],
      "title": "title 4",
      "updated_at": "",
      "user_id": -1,
    },
  },
  "ids": Array [
    4,
    5,
  ],
}
`;

exports[`reducers::notes() responds to events 4`] = `
Object {
  "entities": Object {
    "4": Object {
      "archived": false,
      "body": "body",
      "created_at": "",
      "id": 4,
      "parent_note_id": 0,
      "pinned": false,
      "tags": Array [],
      "title": "title 3",
      "updated_at": "",
      "user_id": -1,
    },
  },
  "ids": Array [
    4,
  ],
}
`;

exports[`reducers::notes() responds to events 5`] = `
Object {
  "entities": Object {},
  "ids": Array [],
}
`;
